Clarification about async access (#124)
OfficeDev/office-scripts-docs: docs/develop/excel-async-model.md

This article will teach you how to write scripts using the legacy, async, APIs. These APIs have the same core functionality as the standard, synchronous Office Scripts APIs, but they require that your script control the data synchronization between the script and the workbook.

> [!IMPORTANT]

> > The async model can only be used with scripts created before the implementation of the current [API model](scripting-fundamentals.md?view=office-scripts). Scripts are permanently locked to the API model they have upon creation. This also means that if you want to convert a legacy script to the new model, you must use a brand new script.

Reviewer: Might still be worth mentioning async is harder - or making a recommendation to transition async to sync if that would be helpful to people? (Not sure)

Author: Yeah, good call.

## `main` function

> Scripts that use the async APIs have a different `main` function. It is an `async` function that has an `Excel.RequestContext` as the first parameter.

Reviewer: ```suggestion
Scripts that use the async APIs have a different `main` function. It's an `async` function that has an `Excel.RequestContext` as the first parameter.
```

}
```

## Converting legacy async scripts to the current model

> The current API model doesn't use `load`, `sync`, or a `RequestContext`. This makes the scripts much easier to write and maintain. Your best resource for converting old scripts is [Stack Overflow](https://stackoverflow.com/questions/tagged/office-scripts). There, you can ask the community for help with specific scenarios. The following guidance should help outline the general steps you'll need to take.

Reviewer: Maybe "current recommended model"?

Author: The async model is solely legacy (unless something changes). Since the only way to make an async script is to copy an old one, it's less of a recommendation and more of the only thing unless you're actively trying to not switch.

> 1. Create a new script and copy the old code into it. Be sure not to include the old `main` method signature, using the current `function main(workbook: ExcelScript.Workbook)` instead.

Reviewer: ```suggestion
1. Create a new script and copy the old async code into it. Be sure not to include the old `main` method signature, using the current `function main(workbook: ExcelScript.Workbook)` instead.
```

2. Remove all the `load` and `sync` calls. They are no longer necessary.

> 3. All properties have been removed. You now access those objects through `get` and `set` methods. For example, instead of setting a cell's fill color through property access like this: `mySheet.getRange("A2:C2").format.fill.color = "blue";`, you'll now use methods like this: `mySheet.getRange("A2:C2").getFormat().getFill().setColor("blue");`

Reviewer: Do you need to say the obvious of "now convert all your other stuff to use get/set"? You imply that here but don't actually say it specifically. Up to you!

Author: I thought about that and since you're thinking that too, I'll add an explicit bit.

4. Collection classes have been replaced by arrays. The `add` and `get` methods of those collection classes were moved to the object that owned the collection, so your references must be updated accordingly. For example, to get a chart named "MyChart" from the first worksheet in the workbook, use the following code: `workbook.getWorksheets()[0].getChart("MyChart");`. Note the `[0]` to access the first value of the `Worksheet[]` returned by `getWorksheets()`.

> 5. Some methods have been renamed for clarity and added for convenience. Please consult the [Office Scripts API reference](/javascript/api/office-scripts/overview?view=office-scripts) for more details.

Reviewer: Is there a plan to do anything to add mapping for people in the reference? Might be helpful

Author: No, but this is fewer than a dozen methods that we have no evidence people are actually using in scripts. I'm not sure there's a record of the changes, so I can't justify the investigation cost.
